### DISCUSSION TITLE
In the driver version after nvidia 410, the Qt test program, when the Qimage size is 816x600, the GPU memory usage is 13M, but when the Qimage size is 817x600, the GPU memory usage continues to increase to 431M

### DISCUSSION THREAD

**Discussion by zhangboin** [NVIDIA/open-gpu-kernel-modules#744] [Q&A] [answered]
In the driver version after nvidia 410, the Qt test program, when the Qimage size is 816x600, the GPU memory usage is 13M, but when the Qimage size is 817x600, the GPU memory usage continues to increase to 431M：
Qt source as follows：
#include <QApplication>
#include <QOpenGLWidget>
#include <QPainter>
#include <QOpenGLContext>
#include <QImage>

class Widget : public QOpenGLWidget
{
public:
    Widget(int width, int height) : QOpenGLWidget(0), image(width, height, QImage::Format_RGB32) {}
    //Widget(int width, int height) : QOpenGLWidget(0), image(width, height, QImage::Format_RGB16) {}
protected:
    void paintGL() override
    {
        glFinish();
        QPainter p(this);
        static unsigned int index = 100;
        image.fill(index++);
        p.drawImage(QPoint(0, 0), image);
        update();
    }
private:
    QImage image;
};

int main(int argc, char** argv)
{
    QApplication app(argc, argv);

    int width = 1024;
    int height = 768;

    if (argc == 3) {
        width = atoi(argv[1]);
        height = atoi(argv[2]);
    }

    Widget w(width, height);
    w.resize(width, height);
    w.show();

    return app.exec();
}
Open the video card driver debug switch, 817x600 will have more ioctl, as follows:

[6714.002807] NVRM: ioctl(0x2a, 0x9870760, 0x20)

Some of the ioctl additions are as follows:

[6714.004671] NVRM: ioctl(0x4a, 0x9872690, 0xb8)

[6714.004870] NVRM: ioctl(0x4a, 0x9872520, 0xb8)

[6714.004921] NVRM: ioctl(0x57, 0x9872400, 0x38)

+++++++++++++++++

[6714.005669] NVRM: ioctl(0x52, 0x9872a20, 0x10)

**Comment by mtijanic** [comment]
The OpenGL driver allocates larger chunks of memory and then sub-allocates from there for texture memory. I guess that this is where you hit a threshold and it decides to allocate another chunk. I imagine if you keep increasing the size after that, the memory usage will stay constant for a while before it makes another jump.

**Reply by zhangboin** [reply]
More test data as follows: （1050tI and 1660S card are same）

When the Qimage size is 1024x1024, the video memory usage is 23 MB
When the Qimage size is 1024x2048, the video memory usage is 43 MB
When the Qimage size is 2048x2048, the video memory usage: 83M
When the Qimage size is 1000x1000, the video memory usage is 23 MB
When the Qimage size is 1024x1025, the video memory usage is 441M

**Reply by mtijanic** [reply]
Heh, there is some weird effect where 2048x2048 is less than 1024x1025, could be due to no power-of-two that triggers some edge case. I'll forward this to the OGL devs to see if there's anything out of the ordinary, but I wouldn't read too much into that info.

By the way, how are you measuring the GPU memory usage? Is this just that app's usage or the total system usage (e.g. as reported by nvidia-smi)

**Reply by zhangboin** [reply]
it reports by nvidia-smi

**Reply by zhangboin** [reply]
However, when the GPU memory is not enough, when the new program is used to apply for video memory, the video memory occupied by the qt test will be dynamically released, so I think it will not be nvidia driver allocated video memory has a lot of cache。
![image](https://github.com/user-attachments/assets/29406722-1dd1-4916-9bd6-498174b23044)

**Reply by zhangboin** [reply]
@mtijanic Hi，are there any other ideas or experimental verification measures for this problem. Thank you!

**Reply by ahuillet** [reply]
This is tracked as NVIDIA bug 5003509 and being worked on.

**Reply by zhangboin** [reply]
Well. thank you. Please let us know if this bug is resolved. And, where can I get a link to the bug 5003509 page.

**Reply by zhangboin** [reply]
@mtijanic @ahuillet Hello, may I ask if this issue has been fixed on 580.65.06？

**Comment by ahuillet** [comment, marked as answer]
Yes, 580.65.06 is expected to carry the fix. (580+ release series has it)

**Reply by zhangboin** [reply]
Thank you !
